language_service/src/lib.rs

```diff
@@ -336,24 +397,25 @@ impl<'a> LanguageService<'a> {
                     continue;
                 }
 
-                self.publish_diagnostics_for_doc(&uri, errors);
+                self.publish_diagnostics_for_doc(&uri, errors).await;
             }
         }
 
         // Clear errors from any documents that previously had errors
         for uri in last_docs_with_errors.difference(&self.documents_with_errors) {
-            self.publish_diagnostics_for_doc(uri, vec![]);
+            self.publish_diagnostics_for_doc(uri, vec![]).await;
         }
     }
 
-    fn publish_diagnostics_for_doc(&self, uri: &str, errors: Vec<Error>) {
+    async fn publish_diagnostics_for_doc(&self, uri: &str, errors: Vec<Error>) {
         let version = self.open_documents.get(uri).map(|d| d.version);
         trace!("publishing diagnostics for {uri} {version:?}): {errors:?}");
         (self.diagnostics_receiver)(DiagnosticUpdate {
             uri: uri.into(),
             version,
             errors,
-        });
+        })
+        .await;
```

Comment:
Was there a reason to make the diagnostics callback async, or is it just for overall consistency? This is essentially just an event - we don't care what the receiver does with it and there is no return value. There is no reason to await it. Unless you needed it for some reason I'd rather it stay a synchronous function.

Reply:
Still, as far as the _language service_ is concerned, it doesn't care about the async-ness. The client can queue up a bunch of work or do whatever it wants in the background. We shouldn't need to await its results. If we do, a bunch of things in the LS get gated behind work we don't care about the results of - we can do better than that.

Did this cause issues in the product or just the tests? If it's a test-only concern I think we can address it in the tests itself (I can actually look into this, since I wrote the diagnostics callbacks in the tests, which I remember being a tearing-my-hair-out process to get the synchronization just right).

Reply:
> This was primarily for test failures. You're suggesting we just call the promise but never await it, and let it execute when the event loop gets around to it?

Yep that's what I'm suggesting.